packages/@aws-cdk/core-synthesizer/lib/synthesizer.ts

```diff
@@ -0,0 +1,79 @@
+import { App, AssetManifestBuilder, DockerImageAssetLocation, DockerImageAssetSource, FileAssetLocation, FileAssetSource, IBoundStackSynthesizer, IReusableStackSynthesizer, ISynthesisSession, Stack, StackSynthesizer } from '@aws-cdk/core';
+import { assertBound } from '@aws-cdk/core/lib/stack-synthesizers/_shared';
+import { IStagingStack, StagingStack } from './staging-stack';
+
+/**
+ * New stack synthesizer properties
+ */
+export interface NewStackSynthesizerProps {
+  /**
+   * Bring a custom staging stack into the app.
+   *
+   * @default - default staging stack
+   */
+  readonly stagingStack?: IStagingStack;
+}
+
+/**
+ * New Stack Synthesizer
+ */
+export class NewStackSynthesizer extends StackSynthesizer implements IReusableStackSynthesizer, IBoundStackSynthesizer {
+  private qualifier?: string;
+  private stagingStack: IStagingStack;
+  private assetManifest = new AssetManifestBuilder();
+
+  constructor(props: NewStackSynthesizerProps = {}) {
+    super();
+
+    // TODO: ensure no tokens
+
+    // TODO: revisit scope
+    this.stagingStack = props.stagingStack ?? new StagingStack(new App(), 'StagingStack', {});
+  }
+
+  /**
+   * // TODO
+   */
+  public reusableBind(stack: Stack): IBoundStackSynthesizer {
+    // Create a copy of the current object and bind that
+    const copy = Object.create(this);
+    copy.bind(stack);
+    return copy;
+  }
+
+  /**
+   * // TODO
+   */
+  public synthesize(_session: ISynthesisSession): void {
+    assertBound(this.qualifier);
+
+    // TODO: finish implementing
+  }
+
+  /**
+   * // TODO
+   */
+  public addFileAsset(asset: FileAssetSource): FileAssetLocation {
+    assertBound(this.stagingStack.stagingBucket.bucketName);
```

Comment:
At this point, we probably need to forward part of the call to the StagingStack, and get the bucket name and object key back.

See [the law of Demeter](https://www.infoworld.com/article/3136224/demystifying-the-law-of-demeter-principle.html): tell, don't ask.